lib/internal/query-executioner.js

```diff
@@ -0,0 +1,55 @@
+const _debugQuery = require('debug')('knex:query');
+const debugBindings = require('debug')('knex:bindings');
+const debugQuery = (sql, txId) => _debugQuery(sql.replace(/%/g, '%%'), txId);
+
+function formatQuery(sql, bindings, timeZone, client) {
+  bindings = bindings == null ? [] : [].concat(bindings);
+  let index = 0;
+  return sql.replace(/\\?\?/g, (match) => {
+    if (match === '\\?') {
+      return '?';
+    }
+    if (index === bindings.length) {
+      return match;
+    }
+    const value = bindings[index++];
+    return client._escapeBinding(value, { timeZone });
+  });
+}
+
+function enrichQueryObject(connection, queryObject, client) {
+  if (typeof queryObject === 'string') {
+    queryObject = { sql: queryObject };
+  }
+
+  queryObject.bindings = client.prepBindings(queryObject.bindings);
+  queryObject.sql = client.positionBindings(queryObject.sql);
+
+  const { __knexUid, __knexTxId } = connection;
+
+  client.emit('query', Object.assign({ __knexUid, __knexTxId }, queryObject));
```

Comment:
I'm not too happy about event emitting and debug logic sneakily being in a method called `enrichQueryObject`, but it is shared across both cases where we use this method.
If someone feels strongly that it's worth to extract those four lines into a separate reused method and has a good name in mind, I'm open to moving them.